layernorm

Krovatkin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63437
* __->__ #63436

use MKLDNN layernorm

use mkldnn version 2

address Elias feedback

fix build CI errors

Differential Revision: [D30388825](https://our.internmc.facebook.com/intern/diff/D30388825)

Krovatkin committed:
layernorm

use MKLDNN layernorm

use mkldnn version 2

address Elias feedback

fix build CI errors

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -271,6 +274,33 @@ Operation createUnaryOp(
+  Tensor bias{};
+  Tensor weight{};
+  auto bias_ival = pop(stack);
+  TORCH_INTERNAL_ASSERT(bias_ival.isTensor());
```
this assertion is already checked in `toTensor` fwiw. although a worse error msg i guess

eellison commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -271,6 +274,33 @@ Operation createUnaryOp(
+  pop(stack);
+  auto eps = pop(stack).toDouble();
+
+  Tensor bias{};
```
nit: `Tensor bias{}` is pretty weird syntax, and takes me a bit to parse reading through the code. Can you just move the initialization down to `Tensor bias = bias_ival.toTensor()` ? (same for weight)

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1927,6 +1927,30 @@ def test_pool3d_batchnorm(self):
+                    if layernorm.elementwise_affine and len(param[1]) == 1:
+                        FileCheck().check("prim::MKLDNNLayerNorm").check_count("aten::to_dense", 1, exactly=True).run(mod.graph)
+                    else:
+                        FileCheck().check_count("aten::to_dense", 1, exactly=True).run(mod.graph)
```
i think i commented this on other pr but this isnt actually checking anything, since we expect to_dense in both if branches. you should check if aten::layernorm exists

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1927,6 +1927,30 @@ def test_pool3d_batchnorm(self):
+    @skipIfNoTorchVision
+    def test_layernorm(self):
+        with set_default_dtype(torch.float):
+            model = torchvision.models.resnet18()
```
idk if we should be instantiating resnet just to get its submodules... but i guess im guilty of that too lol

eellison reviewed (approved):
nice

Krovatkin commented:
@krovatkin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388825).

Krovatkin committed:
Update on "layernorm"


use MKLDNN layernorm

use mkldnn version 2

address Elias feedback

fix build CI errors

Differential Revision: [D30388825](https://our.internmc.facebook.com/intern/diff/D30388825)

[ghstack-poisoned]

Krovatkin commented:
@krovatkin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388825).
